### PR TITLE
[DEV-1756] Improve columns batching logic to account for aggregation parameters

### PR DESCRIPTION
**Pull request by YS-L**
## Description

This improves the columns batching logic in historical features to account for aggregation parameters so that it is not sensitive to the original ordering of features. The change is to introduce a reordering step in the `split_nodes` function.

The previously measured historical features runtime for [`Features V1`](https://github.com/featurebyte/featurebyte-notebooks/blob/f6e8cd4b352c3550540874e9cdd9001ba47a84f0/notebooks/features/grocery/setup.py#L67) was optimistic because the features were ordered meaningfully in the feature list. The runtime increases from 27m to 42m if the feature list is shuffled randomly. With the changes in this PR, the runtime of the shuffled feature list is 25m.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: tests/fixtures/expected_historical_requests_multiple_batches_feature_set_0.sql [hunk 1]
```diff
@@ -134,14 +32,44 @@ WITH "REQUEST_TABLE_W7200_F3600_BS900_M1800_CUSTOMER_ID" AS (
...
 )
 SELECT
   AGG."__FB_ROW_INDEX_FOR_JOIN",
   AGG."POINT_IN_TIME",
   AGG."CUSTOMER_ID",
-  "_fb_internal_window_w7200_avg_30d0e03bfdc9aa70e3001f8c32a5f82e6f793cbb" AS "a_2h_average",
-  "_fb_internal_window_w172800_avg_30d0e03bfdc9aa70e3001f8c32a5f82e6f793cbb" AS "a_48h_average",
+  "_fb_internal_as_at_count_None_membership_status_None_input_2" AS "asat_feature",
   "_fb_internal_item_count_None_order_id_None_input_1" AS "order_size"
 FROM _FB_AGGREGATED AS AGG
```
Comment (gee-senbong) [inline tests/fixtures/expected_historical_requests_multiple_batches_feature_set_0.sql:75]: Seems like the SQL becomes shorter & more readable than before.

File: featurebyte/query_graph/sql/feature_historical.py [hunk 2]
```diff
@@ -365,33 +366,68 @@ def get_historical_features_expr(
...
+        key = ",".join(parts)
+        return key
+
     result = []
-    for i in range(0, len(nodes), num_features_per_query):
-        current_nodes = nodes[i : i + num_features_per_query]
+    sorted_nodes = sorted(nodes, key=get_sort_key)
+    for i in range(0, len(sorted_nodes), num_features_per_query):
+        current_nodes = sorted_nodes[i : i + num_features_per_query]
         result.append(current_nodes)
```
Comment (gee-senbong) [inline featurebyte/query_graph/sql/feature_historical.py:424]: Would you expect to see performance improvement if we could ensure that same source table should be within the same node group (if the size of the group is less than `num_features_per_query`)?

**Review by gee-senbong — APPROVED** [review]
LGTM. Should we add a change log entry for this PR?

File: featurebyte/query_graph/sql/feature_historical.py (hunk 2 above)
Comment (YS-L) [reply featurebyte/query_graph/sql/feature_historical.py:424]: If the size of the group is less than `num_features_per_query`, all the nodes would have been in the same group already regardless of source table, so I think I'm misunderstanding the question. Could you clarify a bit more?

**Comment by YS-L** [comment]
Regarding the changelog, I think the changes here are effectively part of this entry: https://github.com/featurebyte/featurebyte/blob/b908604e4d2a8858c5e1b63b69a3bbddfa191a5d/.changelog/DEV-1740.yaml#L19

since the above change is not published in any version yet.

**Pull request merged**